### PR TITLE
core(fr): add artifact dependency support

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Adds config and runner support for artifact dependencies, see #12017 for expected usage examples and discussion. I know this is a monster PR, but it truly is the minimal set of artifact dependency support that is validated, typechecked, and tested (there isn't even a single gatherer that's upgraded for dependency support in here! 🤯). I do think it's rather helpful to see it all come together rather than jump between different related PRs by potential different reviewers, but we do have few options if we want to reduce the line count.

- Split the config support from runner support and just break some types and/or add temporary meaningless types
- Land the timespan/snapshot runner test updates separately
- Land the config validation pieces separately (everything would typecheck and work but you could create invalid dependency relationships)

**Related Issues/PRs**
ref #11313 
ref #12017

### CODE REVIEW

File: lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      id: 'Artifact',
+      gatherer: {instance: new Gatherer()},
+      dependencies: {ImageElements: {id: 'Dependency'}},
+    };
+    artifactsById = {
+      Dependency: [],
+    };
+  });
+
+  it('should handle no dependencies', async () => {
```
Comment (adamraine) [inline lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js:30]: Would this be the same as `dependencies: {}`?

File: lighthouse-core/test/fraggle-rock/gather/timespan-runner-test.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  it('should connect to the page and run', async () => {
+    const timespan = await startTimespan({page, config});
+    await timespan.endTimespan();
+    expect(mockDriver.connect).toHaveBeenCalled();
+    expect(mockRunnerRun).toHaveBeenCalled();
+  });
+
+  it('should invoke beforeTimespan', async () => {
+
+  });
```
Comment (adamraine) [inline lighthouse-core/test/fraggle-rock/gather/timespan-runner-test.js:71]: ?

File: lighthouse-core/fraggle-rock/config/config.js [hunk 3]
```diff
@@ -47,6 +53,37 @@ function resolveWorkingCopy(configJSON, context) {
...
+        // Check that the phase relationship is OK too.
+        const validDependency = isValidArtifactDependency(gatherer, dependency.gatherer);
+        if (!validDependency) throwInvalidArtifactDependency(artifact.id, dependencyName);
+
+        return [dependencyName, {id: dependency.id}];
+      }
+    );
+
+    return Object.fromEntries(dependencies);
+  } else {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/fraggle-rock/config/config.js:82]: heh I'm normally all about that, but I hate the `(!(` constructions like `in` and `instanceof` so much 😄 

you've convinced me into it though  👍

File: lighthouse-core/fraggle-rock/gather/runner-helpers.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * @param {LH.Config.ArtifactDefn} artifact
+ * @param {Record<string, LH.Gatherer.PhaseResult>} artifactsById
+ * @return {Promise<LH.Gatherer.FRTransitionalContext<LH.Gatherer.DependencyKey>['dependencies']>}
+ */
+async function collectArtifactDependencies(artifact, artifactsById) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/fraggle-rock/gather/runner-helpers.js:13]: would you prefer to convert the final `await` to a `.then`? 

or asking why we want to convert the artifact dependencies from promises to the resolved artifacts?

File: lighthouse-core/fraggle-rock/gather/timespan-runner.js [hunk 5]
```diff
@@ -21,13 +22,17 @@ async function startTimespan(options) {
 
   const requestedUrl = await options.page.url();
 
-  /** @type {Record<string, Promise<Error|undefined>>} */
+  /** @type {Record<string, Promise<void>>} */
   const artifactErrors = {};
+  const dependencies = {};
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/fraggle-rock/gather/timespan-runner.js:27]: it's possible we eventually allow `beforeTimespan` dependencies on other `beforeTimespan` I suppose, but I don't imagine we'll do that very soon.

prefer an inline `dependencies: {}`?

File: lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js:30]: I'm not sure `dependencies: {}` is a valid type given the way I've setup the constraints, but it is semantically equivalent, yes.

If it is an invalid type, are you thinking a `// @ts-expect-error` test anyway?

File: lighthouse-core/test/fraggle-rock/gather/timespan-runner-test.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/test/fraggle-rock/gather/timespan-runner-test.js:71]: fixed :)

**Review by patrickhulce — COMMENTED** [review]
thanks for the review! :)

File: lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/test/fraggle-rock/gather/runner-helpers-test.js:30]: added one for plugin case 👍

File: lighthouse-core/fraggle-rock/config/config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/fraggle-rock/config/config.js]: NGL I tried `'dependencies' not in gatherer.instance.meta` hoping it was possible like in Python.

File: lighthouse-core/fraggle-rock/gather/runner-helpers.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/fraggle-rock/gather/runner-helpers.js:13]: > or asking why we want to convert the artifact dependencies from promises to the resolved artifacts?

This. I don't really have a suggestion, just trying to understand this.

File: lighthouse-core/fraggle-rock/gather/timespan-runner.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/fraggle-rock/gather/timespan-runner.js:27]: Yeah inline SGTM, I got it confused with the second declaration of `dependencies` below

File: lighthouse-core/fraggle-rock/gather/runner-helpers.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/fraggle-rock/gather/runner-helpers.js:13]: We want to resolve any errors ahead of time so the burden doesn't fall onto every individual gatherer to remap their dependencies errors. We do a similar thing for audits today when their artifacts fail.

https://github.com/GoogleChrome/lighthouse/blob/b292ae06ec9666d2a26c6dd44f835fa4f79bf1bf/lighthouse-core/runner.js#L323-L339

File: lighthouse-core/fraggle-rock/gather/timespan-runner.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/fraggle-rock/gather/timespan-runner.js:27]: gotcha, yeah that SG

**Review by adamraine — APPROVED** [review]

**Pull request merged**